runtime/sam/op/router.go

```diff
@@ -93,6 +94,11 @@ func (r *Router) blocked() bool {
 // after receiving the EOS, it's done will be captured as soon as we unblock
 // all channels.
 func (r *Router) sendEOS(err error) bool {
+	defer func() {
+		if r, ok := r.selector.(expr.Resetter); ok {
+			r.Reset()
+		}
+	}()
```

Comment:
How about embedding expr.Resetter in op.Selector? That feels more straightforward to me.